File: android/broker/src/main/kotlin/com/aegis/broker/NativePolicyCoreGate.kt
package com.aegis.broker

import com.sun.jna.Library
import com.sun.jna.Native

/**
 * 原生策略核心的显式启用门禁。
 *
 * 默认构建不启用原生策略核心，保留经验证的 Kotlin Broker。若构建显式要求原生核心，
 * 则原生库、ABI 与完整性验证必须全部就绪；否则任何导航均失败闭合，绝不静默回退。
 */
fun interface NativePolicyCoreGate {
    fun probe(): NativePolicyCoreGateResult
}

data class NativePolicyCoreGateResult(
    val allowsPlatformBroker: Boolean,
    val denialCode: String? = null,
) {
    companion object {
        fun disabled() = NativePolicyCoreGateResult(allowsPlatformBroker = true)

        fun enabled() = NativePolicyCoreGateResult(allowsPlatformBroker = true)

        fun block(denialCode: String) =
            NativePolicyCoreGateResult(
                allowsPlatformBroker = false,
                denialCode = denialCode,
            )
    }
}

object DefaultNativePolicyCoreGate : NativePolicyCoreGate {
    override fun probe(): NativePolicyCoreGateResult {
        if (!BuildConfig.REQUIRE_NATIVE_POLICY_CORE) return NativePolicyCoreGateResult.disabled()
        val abiVersion =
            try {
                Native
                    .load("aegis_policy_core", NativePolicyCoreAbi::class.java)
                    .aegis_policy_core_abi_version()
            } catch (_: LinkageError) {
                return NativePolicyCoreGateResult.block("native_policy_core_unavailable")
            } catch (_: Exception) {
                return NativePolicyCoreGateResult.block("native_policy_core_probe_failed")
            }
        if (abiVersion != EXPECTED_C_ABI_VERSION) {
            return NativePolicyCoreGateResult.block("native_policy_core_abi_mismatch")
        }
        return NativePolicyCoreGateResult.enabled()
    }
}

/**
 * C ABI 版本单源（全库审计 2026-09-02 收敛）：此前 EXPECTED_ABI_VERSION 在
 * NativePolicyCoreGate 与 NativePolicyCoreBridge 双份定义——ABI 升级时漏改
 * 任一处即出现「门禁通过、桥接失配」（或反之）的静默漂移。v3 新增 Rust
 * 托管的确认登记、批准兑换与拒绝接口。
 */
internal const val EXPECTED_C_ABI_VERSION = 3

// JNI 绑定函数名必须与 C ABI 符号逐字一致（snake_case）——豁免命名规范
@Suppress("ktlint:standard:function-naming")
private interface NativePolicyCoreAbi : Library {
    fun aegis_policy_core_abi_version(): Int
}

File: windows/src/Aegis.Windows.App/Broker/NativePolicyCoreGate.cs
namespace Aegis.Windows.Broker;

using System;
using System.Runtime.InteropServices;

/// <summary>
/// 原生策略核心的显式启用门禁。
/// 默认关闭时保持经验证的 C# Broker 路径；显式启用后，任意库加载或 ABI 探测失败均拒绝副作用，
/// 禁止静默切换到另一套策略实现。
/// </summary>
public static class NativePolicyCoreGate
{
    public const string EnableEnvironmentVariable = "AEGIS_REQUIRE_NATIVE_POLICY_CORE";
    public const uint ExpectedAbiVersion = 1;

    public static NativePolicyCoreGateResult ProbeFromEnvironment()
    {
        if (!string.Equals(Environment.GetEnvironmentVariable(EnableEnvironmentVariable), "1", StringComparison.Ordinal))
            return NativePolicyCoreGateResult.Disabled();

        return ProbeLibrary("aegis_policy_core");
    }

    /// <summary>
    /// 探测指定的策略核心库。仅供启动期门禁和构建制品测试使用；失败信息不包含本机绝对路径。
    /// </summary>
    public static NativePolicyCoreGateResult ProbeLibrary(string libraryNameOrPath)
    {
        if (string.IsNullOrWhiteSpace(libraryNameOrPath))
            return NativePolicyCoreGateResult.Block("native_policy_core_path_invalid");

        if (!NativeLibrary.TryLoad(libraryNameOrPath, out var handle))
            return NativePolicyCoreGateResult.Block("native_policy_core_unavailable");

        try
        {
            if (!NativeLibrary.TryGetExport(handle, "aegis_policy_core_abi_version", out var symbol))
                return NativePolicyCoreGateResult.Block("native_policy_core_abi_symbol_missing");

            var abiVersion = Marshal.GetDelegateForFunctionPointer<AbiVersionDelegate>(symbol)();
            return abiVersion == ExpectedAbiVersion
                ? NativePolicyCoreGateResult.Enabled()
                : NativePolicyCoreGateResult.Block("native_policy_core_abi_mismatch");
        }
        catch (Exception)
        {
            return NativePolicyCoreGateResult.Block("native_policy_core_probe_failed");
        }
        finally
        {
            NativeLibrary.Free(handle);
        }
    }

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate uint AbiVersionDelegate();
}

/// <summary>原生策略核心门禁的可审计结果；错误码不包含 URL、令牌或网页内容。</summary>
public sealed record NativePolicyCoreGateResult(bool AllowsPlatformBroker, string? DenialCode)
{
    public static NativePolicyCoreGateResult Disabled() => new(true, null);

    public static NativePolicyCoreGateResult Enabled() => new(true, null);

    public static NativePolicyCoreGateResult Block(string denialCode) => new(false, denialCode);
}
